Tests include all input files.  
Options are already pre-selected for *mPing* TE and this particular test dataset - see [ITIS](https://github.com/Chuan-Jiang/ITIS).  

To run enter:  
`snakemake --use-conda --cores xx`  
`snakemake --cores xx` (when using the singularity container)  

This dataset is taken fro a study investigating transposon polymorphisms using *mPing* TE (insertions result in 3 bp TSDs).  
- genome resequencing project of Japonica A123(SRR631734), which have transposon mping be activted

The test runs the MuWU pipeline on the test dataset and trims the mping consensus sequence from all reads if it is found.  
This way, a subset of reads is created that starts/ends with the TSD region - creating a library suitable for the MuWU insertion identification algorithm.  

Finally, we compare the MuWU found mping insertions with those reported by ITIS.  

In terms of recall we identify all 6 high confidence insertions as reported by ITIS.  
Compared to the RelocaTE2 test, this dataset here is focussed on one particular TE *mPing* - thus we have greater overall TE/TSD read support.  
- The `overlap_support` parameter can be increased to generate higher confidence results and with less false positives.  

Our test shows that in our top 8 MuWU insertion calls we identify the 6 high confidence insertions reported by ITIS.  
This indicates that our approach is well suited for general insertion identification if the input data is suitable - targeted towards one particular TE and with suitable coverage/enrichment.  

Output generated by the test:

```
ITIS finds 6 mping insertions in this dataset.

              V1      V2      V3
1 Chr1:0-2000000  174497  174500
2 Chr1:0-2000000  214352  214355
3 Chr1:0-2000000  316534  316537
4 Chr1:0-2000000  639972  639975
5 Chr1:0-2000000 1193504 1193507
6 Chr1:0-2000000 1374936 1374939
                                                                                 V4
1      SR=8,9,6,2,0,1;MQ=57;NM=mping;GT=8,0:Heter;PV=0.03125;DP=16;TS=1;TE=430;NB=N
2 SR=15,17,9,5,1,2;MQ=57;NM=mping;GT=15,0:Homo;PV=0.00390625;DP=17;TS=1;TE=430;NB=N
3    SR=9,14,3,11,0,0;MQ=57;NM=mping;GT=9,0:Heter;PV=0.03125;DP=20;TS=1;TE=430;NB=N
4   SR=8,8,4,3,0,1;MQ=51;NM=mping;GT=8,14:Heter;PV=0.9903946;DP=28;TS=1;TE=430;NB=N
5     SR=9,12,5,7,0,0;MQ=56;NM=mping;GT=9,0:Heter;PV=0.03125;DP=18;TS=1;TE=430;NB=N
6  SR=8,10,7,2,0,1;MQ=59;NM=mping;GT=8,10:Heter;PV=0.9407654;DP=28;TS=1;TE=430;NB=N
  V5 V6
1  .  +
2  .  +
3  .  -
4  .  +
5  .  +
6  .  +

Overlapping with MuWU results:

           Chr.x Start.x   End.x          Chr.y Start.y   End.y Sample
1 Chr1:0-2000000  174497  174500 Chr1:0-2000000  174498  174500  mping
2 Chr1:0-2000000  214352  214355 Chr1:0-2000000  214353  214355  mping
3 Chr1:0-2000000  316534  316537 Chr1:0-2000000  316535  316537  mping
4 Chr1:0-2000000  639972  639975 Chr1:0-2000000  639973  639975  mping
5 Chr1:0-2000000 1193504 1193507 Chr1:0-2000000 1193505 1193507  mping
6 Chr1:0-2000000 1374936 1374939 Chr1:0-2000000 1374937 1374939  mping
  StartReads EndReads
1          3        6
2          6        6
3          4       13
4          2        5
5          6        7
6          7        8

Consider, that MuWU would normally be used after a targeted PCR approach.
We nevertheless identify 6 of the 6 high confidence mping insertions as identified by ITIS

Also, compared to the test with RelocaTE2 the data test data here is especially enriched for one TE - mping.
Looking at the top supported candidates in our MuWU analysis - StartEnd + EndReads


             Chr   Start     End Sample StartReads EndReads Support
1 Chr1:0-2000000  316535  316537  mping          4       13      17
2 Chr1:0-2000000 1374937 1374939  mping          7        8      15
3 Chr1:0-2000000 1193505 1193507  mping          6        7      13
4 Chr1:0-2000000  214353  214355  mping          6        6      12
5 Chr1:0-2000000  174498  174500  mping          3        6       9
6 Chr1:0-2000000  248110  248112  mping          2        6       8
7 Chr1:0-2000000  352435  352437  mping          3        4       7
8 Chr1:0-2000000  639973  639975  mping          2        5       7

The overlap - 8 topHits MuWU & 6 high confidence candidates from ITIS:

           Chr.x Start.x   End.x          Chr.y Start.y   End.y Sample
1 Chr1:0-2000000  174497  174500 Chr1:0-2000000  174498  174500  mping
2 Chr1:0-2000000  214352  214355 Chr1:0-2000000  214353  214355  mping
3 Chr1:0-2000000  316534  316537 Chr1:0-2000000  316535  316537  mping
4 Chr1:0-2000000  639972  639975 Chr1:0-2000000  639973  639975  mping
5 Chr1:0-2000000 1193504 1193507 Chr1:0-2000000 1193505 1193507  mping
6 Chr1:0-2000000 1374936 1374939 Chr1:0-2000000 1374937 1374939  mping
  StartReads EndReads Support
1          3        6       9
2          6        6      12
3          4       13      17
4          2        5       7
5          6        7      13
6          7        8      15


MuWU overlaps quite well with the results of ITIS.
In the top 8 MuWU candidates we find all 6 ITIS top candidates.
(Note that ITIS and MuWU report insertion positions slightly differently - MuWU reports TSD bp positions.)
```
