Tests include all input files.  
Options are already pre-selected for *mPing* TE and this particular test dataset - see [ITIS](https://github.com/Chuan-Jiang/ITIS).  

To run enter:  
`snakemake --use-conda --cores xx`  
`snakemake --cores xx` (when using the singularity container)  

This data set is taken from genome resequencing project of Japonica A123(SRR631734) investigating transposon polymorphisms using activated *mPing* TE (insertions result in 3 bp TSDs).  

The test runs the MuWU pipeline on the test dataset and trims the *mPing* consensus sequence from all reads if it is found.  
This way, a subset of reads is created that starts/ends with the TSD region - creating a library suitable for the MuWU insertion identification algorithm.  

Finally, we compare the MuWU found *mPing* insertions with those reported by ITIS.  

We identify all 6 high confidence insertions as reported by ITIS.  

Our test shows that in our top 8 MuWU insertion calls we identify all 6 high confidence insertions reported by ITIS.  
This indicates that our approach is well suited for general insertion identification if the input data is suitable - targeted towards one particular TE and with suitable coverage/enrichment.  

Output generated by the test:

```
ITIS finds 6 high-confidence mping insertions in this dataset:

              V1      V2      V3
1 Chr1:0-2000000  174497  174500
2 Chr1:0-2000000  214352  214355
3 Chr1:0-2000000  316534  316537
4 Chr1:0-2000000  639972  639975
5 Chr1:0-2000000 1193504 1193507
6 Chr1:0-2000000 1374936 1374939
                                                                                 V4
1      SR=8,9,6,2,0,1;MQ=57;NM=mping;GT=8,0:Heter;PV=0.03125;DP=16;TS=1;TE=430;NB=N
2 SR=15,17,9,5,1,2;MQ=57;NM=mping;GT=15,0:Homo;PV=0.00390625;DP=17;TS=1;TE=430;NB=N
3    SR=9,14,3,11,0,0;MQ=57;NM=mping;GT=9,0:Heter;PV=0.03125;DP=20;TS=1;TE=430;NB=N
4   SR=8,8,4,3,0,1;MQ=51;NM=mping;GT=8,14:Heter;PV=0.9903946;DP=28;TS=1;TE=430;NB=N
5     SR=9,12,5,7,0,0;MQ=56;NM=mping;GT=9,0:Heter;PV=0.03125;DP=18;TS=1;TE=430;NB=N
6  SR=8,10,7,2,0,1;MQ=59;NM=mping;GT=8,10:Heter;PV=0.9407654;DP=28;TS=1;TE=430;NB=N
  V5 V6
1  .  +
2  .  +
3  .  -
4  .  +
5  .  +
6  .  +

MuWU intersect with ITIS results:

             Chr Start.ITIS End.ITIS Start.MuWU End.MuWU Sample MuWU.StartReads
1 Chr1:0-2000000     174497   174500     174498   174500  mping               3
2 Chr1:0-2000000     214352   214355     214353   214355  mping               6
3 Chr1:0-2000000     316534   316537     316535   316537  mping               4
4 Chr1:0-2000000     639972   639975     639973   639975  mping               2
5 Chr1:0-2000000    1193504  1193507    1193505  1193507  mping               6
6 Chr1:0-2000000    1374936  1374939    1374937  1374939  mping               7
  MuWU.EndReads
1             6
2             6
3            13
4             5
5             7
6             8

Top 8 MuWU insertions contain the 6 high confidence candidates from ITIS:

             Chr   Start     End Sample StartReads EndReads Support
1 Chr1:0-2000000  316535  316537  mping          4       13      17
2 Chr1:0-2000000 1374937 1374939  mping          7        8      15
3 Chr1:0-2000000 1193505 1193507  mping          6        7      13
4 Chr1:0-2000000  214353  214355  mping          6        6      12
5 Chr1:0-2000000  174498  174500  mping          3        6       9
6 Chr1:0-2000000  248110  248112  mping          2        6       8
7 Chr1:0-2000000  352435  352437  mping          3        4       7
8 Chr1:0-2000000  639973  639975  mping          2        5       7
```
