{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3b2b4824",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies ‘pkgconfig’, ‘purrr’, ‘generics’, ‘magrittr’, ‘R6’, ‘tibble’, ‘tidyselect’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "also installing the dependencies ‘lattice’, ‘stringi’, ‘cpp11’, ‘sp’, ‘stringdist’, ‘stringr’, ‘tidyr’, ‘geosphere’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "also installing the dependencies ‘rematch’, ‘hms’, ‘prettyunits’, ‘cellranger’, ‘Rcpp’, ‘progress’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###install.packages(\"tidygenomics\")\n",
    "#install.packages(\"dplyr\")\n",
    "#install.packages(\"fuzzyjoin\")\n",
    "#install.packages(\"readxl\")\n",
    "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#    install.packages(\"BiocManager\")\n",
    "#BiocManager::install(\"IRanges\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a62e46b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#library(tidygenomics)\n",
    "library(dplyr)\n",
    "library(fuzzyjoin)\n",
    "library(readxl)\n",
    "library(IRanges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b5632f30",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "New names:\n",
      "* `` -> ...1\n",
      "\n",
      "`summarise()` has grouped output by 'GeneID'. You can override using the `.groups` argument.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Read in tables with identified germinal Mu insertions; \n",
    "germinal_ins <- read.csv(\"../../results/insertions_table_final/germinal_insertions.csv\", header=TRUE)\n",
    "\n",
    "## Read in stock matrix\n",
    "MY_STOCK <- list.files(\"../../config/stock_matrix/\", pattern=\"\\\\.xlsx$\")\n",
    "stock <- as.data.frame(read_excel(paste0(\"../../config/stock_matrix/\", MY_STOCK)))\n",
    "names(stock)[1] <- \"X__1\"\n",
    "rownames(stock) <- stock$X__1\n",
    "stock <- stock[, -1]\n",
    "stock <- as.matrix(stock)\n",
    "\n",
    "#grid table - to associate arbitrary names with row/col\n",
    "grid_table <- read.csv(\"../../config/grid_sample_sheet.tsv\", sep = \"\\t\", header=TRUE)\n",
    "grid_table <- grid_table %>%\n",
    "                select(dim, base_name)\n",
    "\n",
    "## Read Annotation File\n",
    "#setwd(\"../FGS/\")\n",
    "#MY_gff3 <- list.files(getwd(), pattern=\"\\\\.gff3$\")\n",
    "annotation <- read.delim(\"../../resources/final_annotation_table\", header=FALSE, comment.char=\"#\")\n",
    "# add column names\n",
    "colnames(annotation) <- c(\"GeneID\", \"Chr\", \"Start\", \"End\")\n",
    "\n",
    "# reduce genes to ONE entry which is the longest\n",
    "# multiple entries per gene likely, since after gffread, transcripts of different lengths are included (share name)\n",
    "# we find the min and max positions (start & end coordinate with greatest distance) per gene and keep this record  \n",
    "annotation <- annotation %>%\n",
    "  group_by(GeneID, Chr) %>%\n",
    "  summarise(Start = min(Start), End = max(End))\n",
    "\n",
    "# compute gene_length\n",
    "annotation <- annotation %>% \n",
    "  mutate(Gene_length = End - Start + 1)\n",
    "\n",
    "\n",
    "# relocate Gene column before gene_length\n",
    "#annotation <- annotation %>%\n",
    "#  relocate(Gene, .before = gene_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dd8353ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chr</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>Sample</th><th scope=col>StartReads</th><th scope=col>EndReads</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td> 971410</td><td> 971418</td><td>Col-01</td><td> 2</td><td> 3</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>2113392</td><td>2113400</td><td>Col-01</td><td> 7</td><td>17</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>1</td><td>3772030</td><td>3772038</td><td>Col-01</td><td>11</td><td> 5</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>1</td><td>4340109</td><td>4340117</td><td>Col-01</td><td> 8</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1</td><td>4736284</td><td>4736292</td><td>Col-01</td><td>24</td><td>16</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1</td><td>7558710</td><td>7558718</td><td>Col-01</td><td> 5</td><td>10</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & Chr & Start & End & Sample & StartReads & EndReads\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 1 &  971410 &  971418 & Col-01 &  2 &  3\\\\\n",
       "\t2 & 1 & 2113392 & 2113400 & Col-01 &  7 & 17\\\\\n",
       "\t3 & 1 & 3772030 & 3772038 & Col-01 & 11 &  5\\\\\n",
       "\t4 & 1 & 4340109 & 4340117 & Col-01 &  8 &  2\\\\\n",
       "\t5 & 1 & 4736284 & 4736292 & Col-01 & 24 & 16\\\\\n",
       "\t6 & 1 & 7558710 & 7558718 & Col-01 &  5 & 10\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | Chr &lt;chr&gt; | Start &lt;int&gt; | End &lt;int&gt; | Sample &lt;chr&gt; | StartReads &lt;int&gt; | EndReads &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | 1 |  971410 |  971418 | Col-01 |  2 |  3 |\n",
       "| 2 | 1 | 2113392 | 2113400 | Col-01 |  7 | 17 |\n",
       "| 3 | 1 | 3772030 | 3772038 | Col-01 | 11 |  5 |\n",
       "| 4 | 1 | 4340109 | 4340117 | Col-01 |  8 |  2 |\n",
       "| 5 | 1 | 4736284 | 4736292 | Col-01 | 24 | 16 |\n",
       "| 6 | 1 | 7558710 | 7558718 | Col-01 |  5 | 10 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chr Start   End     Sample StartReads EndReads\n",
       "1 1    971410  971418 Col-01  2          3      \n",
       "2 1   2113392 2113400 Col-01  7         17      \n",
       "3 1   3772030 3772038 Col-01 11          5      \n",
       "4 1   4340109 4340117 Col-01  8          2      \n",
       "5 1   4736284 4736292 Col-01 24         16      \n",
       "6 1   7558710 7558718 Col-01  5         10      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>dim</th><th scope=col>num</th><th scope=col>base_name</th><th scope=col>fq_1_file</th><th scope=col>fq_2_file</th><th scope=col>fq_1_end</th><th scope=col>fq_2_end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>row</td><td>1</td><td>Row-01</td><td>Row-01_1.fq.gz</td><td>Row-01_2.fq.gz</td><td>_1.fq.gz</td><td>_2.fq.gz</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>row</td><td>2</td><td>Row-02</td><td>Row-02_1.fq.gz</td><td>Row-02_2.fq.gz</td><td>_1.fq.gz</td><td>_2.fq.gz</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>row</td><td>3</td><td>Row-03</td><td>Row-03_1.fq.gz</td><td>Row-03_2.fq.gz</td><td>_1.fq.gz</td><td>_2.fq.gz</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>row</td><td>4</td><td>Row-04</td><td>Row-04_1.fq.gz</td><td>Row-04_2.fq.gz</td><td>_1.fq.gz</td><td>_2.fq.gz</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>col</td><td>1</td><td>Col-01</td><td>Col-01_1.fq.gz</td><td>Col-01_2.fq.gz</td><td>_1.fq.gz</td><td>_2.fq.gz</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>col</td><td>2</td><td>Col-02</td><td>Col-02_1.fq.gz</td><td>Col-02_2.fq.gz</td><td>_1.fq.gz</td><td>_2.fq.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & dim & num & base\\_name & fq\\_1\\_file & fq\\_2\\_file & fq\\_1\\_end & fq\\_2\\_end\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & row & 1 & Row-01 & Row-01\\_1.fq.gz & Row-01\\_2.fq.gz & \\_1.fq.gz & \\_2.fq.gz\\\\\n",
       "\t2 & row & 2 & Row-02 & Row-02\\_1.fq.gz & Row-02\\_2.fq.gz & \\_1.fq.gz & \\_2.fq.gz\\\\\n",
       "\t3 & row & 3 & Row-03 & Row-03\\_1.fq.gz & Row-03\\_2.fq.gz & \\_1.fq.gz & \\_2.fq.gz\\\\\n",
       "\t4 & row & 4 & Row-04 & Row-04\\_1.fq.gz & Row-04\\_2.fq.gz & \\_1.fq.gz & \\_2.fq.gz\\\\\n",
       "\t5 & col & 1 & Col-01 & Col-01\\_1.fq.gz & Col-01\\_2.fq.gz & \\_1.fq.gz & \\_2.fq.gz\\\\\n",
       "\t6 & col & 2 & Col-02 & Col-02\\_1.fq.gz & Col-02\\_2.fq.gz & \\_1.fq.gz & \\_2.fq.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | dim &lt;chr&gt; | num &lt;int&gt; | base_name &lt;chr&gt; | fq_1_file &lt;chr&gt; | fq_2_file &lt;chr&gt; | fq_1_end &lt;chr&gt; | fq_2_end &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | row | 1 | Row-01 | Row-01_1.fq.gz | Row-01_2.fq.gz | _1.fq.gz | _2.fq.gz |\n",
       "| 2 | row | 2 | Row-02 | Row-02_1.fq.gz | Row-02_2.fq.gz | _1.fq.gz | _2.fq.gz |\n",
       "| 3 | row | 3 | Row-03 | Row-03_1.fq.gz | Row-03_2.fq.gz | _1.fq.gz | _2.fq.gz |\n",
       "| 4 | row | 4 | Row-04 | Row-04_1.fq.gz | Row-04_2.fq.gz | _1.fq.gz | _2.fq.gz |\n",
       "| 5 | col | 1 | Col-01 | Col-01_1.fq.gz | Col-01_2.fq.gz | _1.fq.gz | _2.fq.gz |\n",
       "| 6 | col | 2 | Col-02 | Col-02_1.fq.gz | Col-02_2.fq.gz | _1.fq.gz | _2.fq.gz |\n",
       "\n"
      ],
      "text/plain": [
       "  dim num base_name fq_1_file      fq_2_file      fq_1_end fq_2_end\n",
       "1 row 1   Row-01    Row-01_1.fq.gz Row-01_2.fq.gz _1.fq.gz _2.fq.gz\n",
       "2 row 2   Row-02    Row-02_1.fq.gz Row-02_2.fq.gz _1.fq.gz _2.fq.gz\n",
       "3 row 3   Row-03    Row-03_1.fq.gz Row-03_2.fq.gz _1.fq.gz _2.fq.gz\n",
       "4 row 4   Row-04    Row-04_1.fq.gz Row-04_2.fq.gz _1.fq.gz _2.fq.gz\n",
       "5 col 1   Col-01    Col-01_1.fq.gz Col-01_2.fq.gz _1.fq.gz _2.fq.gz\n",
       "6 col 2   Col-02    Col-02_1.fq.gz Col-02_2.fq.gz _1.fq.gz _2.fq.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>GeneID</th><th scope=col>Chr</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>Gene_length</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSRNA049437471</td><td>1</td><td> 2815814</td><td> 2815887</td><td>74</td></tr>\n",
       "\t<tr><td>ENSRNA049437473</td><td>1</td><td> 2998702</td><td> 2998773</td><td>72</td></tr>\n",
       "\t<tr><td>ENSRNA049437518</td><td>1</td><td> 5888817</td><td> 5888890</td><td>74</td></tr>\n",
       "\t<tr><td>ENSRNA049437607</td><td>1</td><td>10109832</td><td>10109903</td><td>72</td></tr>\n",
       "\t<tr><td>ENSRNA049437614</td><td>1</td><td>10573120</td><td>10573190</td><td>71</td></tr>\n",
       "\t<tr><td>ENSRNA049437658</td><td>1</td><td>18379795</td><td>18379867</td><td>73</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 6 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " GeneID & Chr & Start & End & Gene\\_length\\\\\n",
       " <chr> & <chr> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t ENSRNA049437471 & 1 &  2815814 &  2815887 & 74\\\\\n",
       "\t ENSRNA049437473 & 1 &  2998702 &  2998773 & 72\\\\\n",
       "\t ENSRNA049437518 & 1 &  5888817 &  5888890 & 74\\\\\n",
       "\t ENSRNA049437607 & 1 & 10109832 & 10109903 & 72\\\\\n",
       "\t ENSRNA049437614 & 1 & 10573120 & 10573190 & 71\\\\\n",
       "\t ENSRNA049437658 & 1 & 18379795 & 18379867 & 73\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 6 × 5\n",
       "\n",
       "| GeneID &lt;chr&gt; | Chr &lt;chr&gt; | Start &lt;int&gt; | End &lt;int&gt; | Gene_length &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| ENSRNA049437471 | 1 |  2815814 |  2815887 | 74 |\n",
       "| ENSRNA049437473 | 1 |  2998702 |  2998773 | 72 |\n",
       "| ENSRNA049437518 | 1 |  5888817 |  5888890 | 74 |\n",
       "| ENSRNA049437607 | 1 | 10109832 | 10109903 | 72 |\n",
       "| ENSRNA049437614 | 1 | 10573120 | 10573190 | 71 |\n",
       "| ENSRNA049437658 | 1 | 18379795 | 18379867 | 73 |\n",
       "\n"
      ],
      "text/plain": [
       "  GeneID          Chr Start    End      Gene_length\n",
       "1 ENSRNA049437471 1    2815814  2815887 74         \n",
       "2 ENSRNA049437473 1    2998702  2998773 72         \n",
       "3 ENSRNA049437518 1    5888817  5888890 74         \n",
       "4 ENSRNA049437607 1   10109832 10109903 72         \n",
       "5 ENSRNA049437614 1   10573120 10573190 71         \n",
       "6 ENSRNA049437658 1   18379795 18379867 73         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "46430"
      ],
      "text/latex": [
       "46430"
      ],
      "text/markdown": [
       "46430"
      ],
      "text/plain": [
       "[1] 46430"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(germinal_ins)\n",
    "head(grid_table)\n",
    "head(annotation)\n",
    "nrow(annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "00bda16a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chr.x</th><th scope=col>Start.x</th><th scope=col>End.x</th><th scope=col>Sample</th><th scope=col>StartReads</th><th scope=col>EndReads</th><th scope=col>GeneID</th><th scope=col>Chr.y</th><th scope=col>Start.y</th><th scope=col>End.y</th><th scope=col>Gene_length</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td> 2113392</td><td> 2113400</td><td>Col-01</td><td> 7</td><td>17</td><td>Zm00001d027296</td><td>1</td><td> 2109189</td><td> 2113478</td><td> 4290</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td> 4736284</td><td> 4736292</td><td>Col-01</td><td>24</td><td>16</td><td>Zm00001d027412</td><td>1</td><td> 4724456</td><td> 4736832</td><td>12377</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>1</td><td> 7558710</td><td> 7558718</td><td>Col-01</td><td> 5</td><td>10</td><td>Zm00001d027532</td><td>1</td><td> 7558685</td><td> 7564087</td><td> 5403</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>1</td><td>15014601</td><td>15014609</td><td>Col-01</td><td> 4</td><td> 2</td><td>Zm00001d027845</td><td>1</td><td>15013950</td><td>15018699</td><td> 4750</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1</td><td>15014626</td><td>15014634</td><td>Col-01</td><td> 2</td><td> 7</td><td>Zm00001d027845</td><td>1</td><td>15013950</td><td>15018699</td><td> 4750</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1</td><td>21998842</td><td>21998850</td><td>Col-01</td><td>23</td><td>13</td><td>Zm00001d028073</td><td>1</td><td>21990180</td><td>22006881</td><td>16702</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Chr.x & Start.x & End.x & Sample & StartReads & EndReads & GeneID & Chr.y & Start.y & End.y & Gene\\_length\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <int> & <int> & <chr> & <chr> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1 &  2113392 &  2113400 & Col-01 &  7 & 17 & Zm00001d027296 & 1 &  2109189 &  2113478 &  4290\\\\\n",
       "\t2 & 1 &  4736284 &  4736292 & Col-01 & 24 & 16 & Zm00001d027412 & 1 &  4724456 &  4736832 & 12377\\\\\n",
       "\t3 & 1 &  7558710 &  7558718 & Col-01 &  5 & 10 & Zm00001d027532 & 1 &  7558685 &  7564087 &  5403\\\\\n",
       "\t4 & 1 & 15014601 & 15014609 & Col-01 &  4 &  2 & Zm00001d027845 & 1 & 15013950 & 15018699 &  4750\\\\\n",
       "\t5 & 1 & 15014626 & 15014634 & Col-01 &  2 &  7 & Zm00001d027845 & 1 & 15013950 & 15018699 &  4750\\\\\n",
       "\t6 & 1 & 21998842 & 21998850 & Col-01 & 23 & 13 & Zm00001d028073 & 1 & 21990180 & 22006881 & 16702\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 11\n",
       "\n",
       "| <!--/--> | Chr.x &lt;chr&gt; | Start.x &lt;int&gt; | End.x &lt;int&gt; | Sample &lt;chr&gt; | StartReads &lt;int&gt; | EndReads &lt;int&gt; | GeneID &lt;chr&gt; | Chr.y &lt;chr&gt; | Start.y &lt;int&gt; | End.y &lt;int&gt; | Gene_length &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 |  2113392 |  2113400 | Col-01 |  7 | 17 | Zm00001d027296 | 1 |  2109189 |  2113478 |  4290 |\n",
       "| 2 | 1 |  4736284 |  4736292 | Col-01 | 24 | 16 | Zm00001d027412 | 1 |  4724456 |  4736832 | 12377 |\n",
       "| 3 | 1 |  7558710 |  7558718 | Col-01 |  5 | 10 | Zm00001d027532 | 1 |  7558685 |  7564087 |  5403 |\n",
       "| 4 | 1 | 15014601 | 15014609 | Col-01 |  4 |  2 | Zm00001d027845 | 1 | 15013950 | 15018699 |  4750 |\n",
       "| 5 | 1 | 15014626 | 15014634 | Col-01 |  2 |  7 | Zm00001d027845 | 1 | 15013950 | 15018699 |  4750 |\n",
       "| 6 | 1 | 21998842 | 21998850 | Col-01 | 23 | 13 | Zm00001d028073 | 1 | 21990180 | 22006881 | 16702 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chr.x Start.x  End.x    Sample StartReads EndReads GeneID         Chr.y\n",
       "1 1      2113392  2113400 Col-01  7         17       Zm00001d027296 1    \n",
       "2 1      4736284  4736292 Col-01 24         16       Zm00001d027412 1    \n",
       "3 1      7558710  7558718 Col-01  5         10       Zm00001d027532 1    \n",
       "4 1     15014601 15014609 Col-01  4          2       Zm00001d027845 1    \n",
       "5 1     15014626 15014634 Col-01  2          7       Zm00001d027845 1    \n",
       "6 1     21998842 21998850 Col-01 23         13       Zm00001d028073 1    \n",
       "  Start.y  End.y    Gene_length\n",
       "1  2109189  2113478  4290      \n",
       "2  4724456  4736832 12377      \n",
       "3  7558685  7564087  5403      \n",
       "4 15013950 15018699  4750      \n",
       "5 15013950 15018699  4750      \n",
       "6 21990180 22006881 16702      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# join MuGerminal table with annotation\n",
    "germinal_ins_annotated <- fuzzyjoin::genome_inner_join(germinal_ins, \n",
    "                                annotation, \n",
    "                                by=c(\"Chr\", \"Start\", \"End\") \n",
    "                                )\n",
    "head(germinal_ins_annotated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0ea8cac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename, relocate and drop the columns to create a useful table\n",
    "germinal_ins_annotated <- germinal_ins_annotated %>%\n",
    "  dplyr::rename(\n",
    "    Chr = Chr.y,\n",
    "    Start = Start.y,\n",
    "    End = End.y,\n",
    "    InsertionStart = Start.x,\n",
    "    InsertionEnd = End.x,\n",
    "  ) %>%\n",
    "  relocate(\n",
    "    Sample, InsertionStart, InsertionEnd, StartReads, EndReads, .before = Gene_length\n",
    "  ) %>%\n",
    "  select(\n",
    "    -Chr.x\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6344995e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GeneID</th><th scope=col>Chr</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>Sample</th><th scope=col>InsertionStart</th><th scope=col>InsertionEnd</th><th scope=col>StartReads</th><th scope=col>EndReads</th><th scope=col>Gene_length</th><th scope=col>stock</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Zm00001d027296</td><td>1</td><td>2109189</td><td>2113478</td><td>Row-01</td><td>2113392</td><td>2113400</td><td>15</td><td>12</td><td> 4290</td><td>D-0078</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Zm00001d027296</td><td>1</td><td>2109189</td><td>2113478</td><td>Col-01</td><td>2113392</td><td>2113400</td><td> 7</td><td>17</td><td> 4290</td><td>D-0078</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Zm00001d027320</td><td>1</td><td>2878535</td><td>2889452</td><td>Row-03</td><td>2889029</td><td>2889037</td><td> 5</td><td> 4</td><td>10918</td><td>D-0129</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Zm00001d027320</td><td>1</td><td>2878535</td><td>2889452</td><td>Col-04</td><td>2889029</td><td>2889037</td><td> 2</td><td> 2</td><td>10918</td><td>D-0129</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Zm00001d027321</td><td>1</td><td>2883539</td><td>2892612</td><td>Row-03</td><td>2889029</td><td>2889037</td><td> 5</td><td> 4</td><td> 9074</td><td>D-0129</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Zm00001d027321</td><td>1</td><td>2883539</td><td>2892612</td><td>Col-04</td><td>2889029</td><td>2889037</td><td> 2</td><td> 2</td><td> 9074</td><td>D-0129</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>Zm00001d027324</td><td>1</td><td>2959502</td><td>2972168</td><td>Row-01</td><td>2959609</td><td>2959617</td><td> 7</td><td>11</td><td>12667</td><td>D-0081</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>Zm00001d027324</td><td>1</td><td>2959502</td><td>2972168</td><td>Col-04</td><td>2959609</td><td>2959617</td><td>25</td><td>72</td><td>12667</td><td>D-0081</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>Zm00001d027344</td><td>1</td><td>3371324</td><td>3374866</td><td>Row-02</td><td>3372136</td><td>3372144</td><td>11</td><td> 8</td><td> 3543</td><td>D-0104</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>Zm00001d027344</td><td>1</td><td>3371324</td><td>3374866</td><td>Col-03</td><td>3372136</td><td>3372144</td><td> 6</td><td>12</td><td> 3543</td><td>D-0104</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & GeneID & Chr & Start & End & Sample & InsertionStart & InsertionEnd & StartReads & EndReads & Gene\\_length & stock\\\\\n",
       "  & <chr> & <chr> & <int> & <int> & <chr> & <int> & <int> & <int> & <int> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & Zm00001d027296 & 1 & 2109189 & 2113478 & Row-01 & 2113392 & 2113400 & 15 & 12 &  4290 & D-0078\\\\\n",
       "\t2 & Zm00001d027296 & 1 & 2109189 & 2113478 & Col-01 & 2113392 & 2113400 &  7 & 17 &  4290 & D-0078\\\\\n",
       "\t3 & Zm00001d027320 & 1 & 2878535 & 2889452 & Row-03 & 2889029 & 2889037 &  5 &  4 & 10918 & D-0129\\\\\n",
       "\t4 & Zm00001d027320 & 1 & 2878535 & 2889452 & Col-04 & 2889029 & 2889037 &  2 &  2 & 10918 & D-0129\\\\\n",
       "\t5 & Zm00001d027321 & 1 & 2883539 & 2892612 & Row-03 & 2889029 & 2889037 &  5 &  4 &  9074 & D-0129\\\\\n",
       "\t6 & Zm00001d027321 & 1 & 2883539 & 2892612 & Col-04 & 2889029 & 2889037 &  2 &  2 &  9074 & D-0129\\\\\n",
       "\t7 & Zm00001d027324 & 1 & 2959502 & 2972168 & Row-01 & 2959609 & 2959617 &  7 & 11 & 12667 & D-0081\\\\\n",
       "\t8 & Zm00001d027324 & 1 & 2959502 & 2972168 & Col-04 & 2959609 & 2959617 & 25 & 72 & 12667 & D-0081\\\\\n",
       "\t9 & Zm00001d027344 & 1 & 3371324 & 3374866 & Row-02 & 3372136 & 3372144 & 11 &  8 &  3543 & D-0104\\\\\n",
       "\t10 & Zm00001d027344 & 1 & 3371324 & 3374866 & Col-03 & 3372136 & 3372144 &  6 & 12 &  3543 & D-0104\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 11\n",
       "\n",
       "| <!--/--> | GeneID &lt;chr&gt; | Chr &lt;chr&gt; | Start &lt;int&gt; | End &lt;int&gt; | Sample &lt;chr&gt; | InsertionStart &lt;int&gt; | InsertionEnd &lt;int&gt; | StartReads &lt;int&gt; | EndReads &lt;int&gt; | Gene_length &lt;dbl&gt; | stock &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Zm00001d027296 | 1 | 2109189 | 2113478 | Row-01 | 2113392 | 2113400 | 15 | 12 |  4290 | D-0078 |\n",
       "| 2 | Zm00001d027296 | 1 | 2109189 | 2113478 | Col-01 | 2113392 | 2113400 |  7 | 17 |  4290 | D-0078 |\n",
       "| 3 | Zm00001d027320 | 1 | 2878535 | 2889452 | Row-03 | 2889029 | 2889037 |  5 |  4 | 10918 | D-0129 |\n",
       "| 4 | Zm00001d027320 | 1 | 2878535 | 2889452 | Col-04 | 2889029 | 2889037 |  2 |  2 | 10918 | D-0129 |\n",
       "| 5 | Zm00001d027321 | 1 | 2883539 | 2892612 | Row-03 | 2889029 | 2889037 |  5 |  4 |  9074 | D-0129 |\n",
       "| 6 | Zm00001d027321 | 1 | 2883539 | 2892612 | Col-04 | 2889029 | 2889037 |  2 |  2 |  9074 | D-0129 |\n",
       "| 7 | Zm00001d027324 | 1 | 2959502 | 2972168 | Row-01 | 2959609 | 2959617 |  7 | 11 | 12667 | D-0081 |\n",
       "| 8 | Zm00001d027324 | 1 | 2959502 | 2972168 | Col-04 | 2959609 | 2959617 | 25 | 72 | 12667 | D-0081 |\n",
       "| 9 | Zm00001d027344 | 1 | 3371324 | 3374866 | Row-02 | 3372136 | 3372144 | 11 |  8 |  3543 | D-0104 |\n",
       "| 10 | Zm00001d027344 | 1 | 3371324 | 3374866 | Col-03 | 3372136 | 3372144 |  6 | 12 |  3543 | D-0104 |\n",
       "\n"
      ],
      "text/plain": [
       "   GeneID         Chr Start   End     Sample InsertionStart InsertionEnd\n",
       "1  Zm00001d027296 1   2109189 2113478 Row-01 2113392        2113400     \n",
       "2  Zm00001d027296 1   2109189 2113478 Col-01 2113392        2113400     \n",
       "3  Zm00001d027320 1   2878535 2889452 Row-03 2889029        2889037     \n",
       "4  Zm00001d027320 1   2878535 2889452 Col-04 2889029        2889037     \n",
       "5  Zm00001d027321 1   2883539 2892612 Row-03 2889029        2889037     \n",
       "6  Zm00001d027321 1   2883539 2892612 Col-04 2889029        2889037     \n",
       "7  Zm00001d027324 1   2959502 2972168 Row-01 2959609        2959617     \n",
       "8  Zm00001d027324 1   2959502 2972168 Col-04 2959609        2959617     \n",
       "9  Zm00001d027344 1   3371324 3374866 Row-02 3372136        3372144     \n",
       "10 Zm00001d027344 1   3371324 3374866 Col-03 3372136        3372144     \n",
       "   StartReads EndReads Gene_length stock \n",
       "1  15         12        4290       D-0078\n",
       "2   7         17        4290       D-0078\n",
       "3   5          4       10918       D-0129\n",
       "4   2          2       10918       D-0129\n",
       "5   5          4        9074       D-0129\n",
       "6   2          2        9074       D-0129\n",
       "7   7         11       12667       D-0081\n",
       "8  25         72       12667       D-0081\n",
       "9  11          8        3543       D-0104\n",
       "10  6         12        3543       D-0104"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "germinal_ins_grid_info <- merge(germinal_ins_annotated, grid_table, by.x = \"Sample\", by.y = \"base_name\")\n",
    "\n",
    "# order the table as to have intersecting samples next to one another\n",
    "germinal_ins_grid_info <- germinal_ins_grid_info %>%\n",
    "  arrange(., Chr, Start, InsertionStart)\n",
    "\n",
    "\n",
    "## assigning Stock\n",
    "#create seperate dataframes - just with Cols and just with Rows\n",
    "just_row <- germinal_ins_grid_info %>%\n",
    "  filter(dim == \"row\") %>%\n",
    "  select(-dim)\n",
    "  \n",
    "just_col <- germinal_ins_grid_info %>%\n",
    "  filter(dim == \"col\") %>%\n",
    "  select(-dim)\n",
    "\n",
    "full <- c()\n",
    "for (i in 1:nrow(just_row))  {\n",
    "  full[i] <- stock[cbind(just_row$Sample[i], just_col$Sample[i])]\n",
    "}\n",
    "\n",
    "just_col$stock <- full\n",
    "just_row$stock <- full\n",
    "\n",
    "germinal_ins_annotated_stocks <- rbind(just_row, just_col)\n",
    "\n",
    "# order the table as to have intersecting samples next to one another; again..\n",
    "germinal_ins_annotated_stocks <- germinal_ins_annotated_stocks %>%\n",
    "  arrange(., Chr, Start, InsertionStart) %>%\n",
    "  relocate(Sample, .before = InsertionStart)\n",
    "\n",
    "\n",
    "head(germinal_ins_annotated_stocks, n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "84a46fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#setwd(\"MuSeq_table_final/\")\n",
    "write.csv(germinal_ins_annotated_stocks, \n",
    "          \"../../results/insertions_table_final/germinal_identified_insertions_annotated.csv\", \n",
    "          row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6f456e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1516512b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "ef95d4f6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "2334"
      ],
      "text/latex": [
       "2334"
      ],
      "text/markdown": [
       "2334"
      ],
      "text/plain": [
       "[1] 2334"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "2254"
      ],
      "text/latex": [
       "2254"
      ],
      "text/markdown": [
       "2254"
      ],
      "text/plain": [
       "[1] 2254"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GeneID</th><th scope=col>Chromosome</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>Sample</th><th scope=col>StartReads</th><th scope=col>EndReads</th><th scope=col>Gene_length</th><th scope=col>stock</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSRNA049997441-T1</td><td>7           </td><td>39887977</td><td>39887985</td><td>Row-01</td><td> 8</td><td> 9</td><td>  NA</td><td>D-0081</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSRNA049997441-T1</td><td>7           </td><td>39887977</td><td>39887985</td><td>Col-04</td><td>28</td><td>39</td><td>  NA</td><td>D-0081</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Zm00001d000035    </td><td>B73V4_ctg182</td><td>  387130</td><td>  387138</td><td>Row-01</td><td> 8</td><td> 7</td><td>2873</td><td>D-0080</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Zm00001d000035    </td><td>B73V4_ctg182</td><td>  387130</td><td>  387138</td><td>Col-03</td><td>30</td><td> 9</td><td>2873</td><td>D-0080</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Zm00001d000055    </td><td>B73V4_ctg10 </td><td>   97290</td><td>   97298</td><td>Row-01</td><td>12</td><td>29</td><td>3357</td><td>D-0079</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Zm00001d000055    </td><td>B73V4_ctg10 </td><td>   97290</td><td>   97298</td><td>Col-02</td><td>22</td><td>15</td><td>3357</td><td>D-0079</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & GeneID & Chromosome & Start & End & Sample & StartReads & EndReads & Gene\\_length & stock\\\\\n",
       "  & <chr> & <chr> & <int> & <int> & <chr> & <int> & <int> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSRNA049997441-T1 & 7            & 39887977 & 39887985 & Row-01 &  8 &  9 &   NA & D-0081\\\\\n",
       "\t2 & ENSRNA049997441-T1 & 7            & 39887977 & 39887985 & Col-04 & 28 & 39 &   NA & D-0081\\\\\n",
       "\t3 & Zm00001d000035     & B73V4\\_ctg182 &   387130 &   387138 & Row-01 &  8 &  7 & 2873 & D-0080\\\\\n",
       "\t4 & Zm00001d000035     & B73V4\\_ctg182 &   387130 &   387138 & Col-03 & 30 &  9 & 2873 & D-0080\\\\\n",
       "\t5 & Zm00001d000055     & B73V4\\_ctg10  &    97290 &    97298 & Row-01 & 12 & 29 & 3357 & D-0079\\\\\n",
       "\t6 & Zm00001d000055     & B73V4\\_ctg10  &    97290 &    97298 & Col-02 & 22 & 15 & 3357 & D-0079\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | GeneID &lt;chr&gt; | Chromosome &lt;chr&gt; | Start &lt;int&gt; | End &lt;int&gt; | Sample &lt;chr&gt; | StartReads &lt;int&gt; | EndReads &lt;int&gt; | Gene_length &lt;int&gt; | stock &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSRNA049997441-T1 | 7            | 39887977 | 39887985 | Row-01 |  8 |  9 |   NA | D-0081 |\n",
       "| 2 | ENSRNA049997441-T1 | 7            | 39887977 | 39887985 | Col-04 | 28 | 39 |   NA | D-0081 |\n",
       "| 3 | Zm00001d000035     | B73V4_ctg182 |   387130 |   387138 | Row-01 |  8 |  7 | 2873 | D-0080 |\n",
       "| 4 | Zm00001d000035     | B73V4_ctg182 |   387130 |   387138 | Col-03 | 30 |  9 | 2873 | D-0080 |\n",
       "| 5 | Zm00001d000055     | B73V4_ctg10  |    97290 |    97298 | Row-01 | 12 | 29 | 3357 | D-0079 |\n",
       "| 6 | Zm00001d000055     | B73V4_ctg10  |    97290 |    97298 | Col-02 | 22 | 15 | 3357 | D-0079 |\n",
       "\n"
      ],
      "text/plain": [
       "  GeneID             Chromosome   Start    End      Sample StartReads EndReads\n",
       "1 ENSRNA049997441-T1 7            39887977 39887985 Row-01  8          9      \n",
       "2 ENSRNA049997441-T1 7            39887977 39887985 Col-04 28         39      \n",
       "3 Zm00001d000035     B73V4_ctg182   387130   387138 Row-01  8          7      \n",
       "4 Zm00001d000035     B73V4_ctg182   387130   387138 Col-03 30          9      \n",
       "5 Zm00001d000055     B73V4_ctg10     97290    97298 Row-01 12         29      \n",
       "6 Zm00001d000055     B73V4_ctg10     97290    97298 Col-02 22         15      \n",
       "  Gene_length stock \n",
       "1   NA        D-0081\n",
       "2   NA        D-0081\n",
       "3 2873        D-0080\n",
       "4 2873        D-0080\n",
       "5 3357        D-0079\n",
       "6 3357        D-0079"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>GeneID</th><th scope=col>Chromosome</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>Sample</th><th scope=col>StartReads</th><th scope=col>EndReads</th><th scope=col>Gene_length</th><th scope=col>stock</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSRNA049997441-T1</td><td>7</td><td>39887977</td><td>39887985</td><td>Row-01</td><td> 8</td><td> 9</td><td>NA</td><td>D-0081</td></tr>\n",
       "\t<tr><td>ENSRNA049997441-T1</td><td>7</td><td>39887977</td><td>39887985</td><td>Col-04</td><td>28</td><td>39</td><td>NA</td><td>D-0081</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " GeneID & Chromosome & Start & End & Sample & StartReads & EndReads & Gene\\_length & stock\\\\\n",
       " <chr> & <chr> & <int> & <int> & <chr> & <int> & <int> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t ENSRNA049997441-T1 & 7 & 39887977 & 39887985 & Row-01 &  8 &  9 & NA & D-0081\\\\\n",
       "\t ENSRNA049997441-T1 & 7 & 39887977 & 39887985 & Col-04 & 28 & 39 & NA & D-0081\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| GeneID &lt;chr&gt; | Chromosome &lt;chr&gt; | Start &lt;int&gt; | End &lt;int&gt; | Sample &lt;chr&gt; | StartReads &lt;int&gt; | EndReads &lt;int&gt; | Gene_length &lt;int&gt; | stock &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| ENSRNA049997441-T1 | 7 | 39887977 | 39887985 | Row-01 |  8 |  9 | NA | D-0081 |\n",
       "| ENSRNA049997441-T1 | 7 | 39887977 | 39887985 | Col-04 | 28 | 39 | NA | D-0081 |\n",
       "\n"
      ],
      "text/plain": [
       "  GeneID             Chromosome Start    End      Sample StartReads EndReads\n",
       "1 ENSRNA049997441-T1 7          39887977 39887985 Row-01  8          9      \n",
       "2 ENSRNA049997441-T1 7          39887977 39887985 Col-04 28         39      \n",
       "  Gene_length stock \n",
       "1 NA          D-0081\n",
       "2 NA          D-0081"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 80 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>GeneID</th><th scope=col>Chr</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>Sample</th><th scope=col>InsertionStart</th><th scope=col>InsertionEnd</th><th scope=col>StartReads</th><th scope=col>EndReads</th><th scope=col>Gene_length</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Zm00001d027801</td><td>1 </td><td> 13815367</td><td> 13827552</td><td>Col-02</td><td> 13815363</td><td> 13815371</td><td> 6</td><td> 9</td><td>12186</td></tr>\n",
       "\t<tr><td>Zm00001d027801</td><td>1 </td><td> 13815367</td><td> 13827552</td><td>Row-01</td><td> 13815363</td><td> 13815371</td><td> 6</td><td> 4</td><td>12186</td></tr>\n",
       "\t<tr><td>Zm00001d028818</td><td>1 </td><td> 47566046</td><td> 47589502</td><td>Col-01</td><td> 47566043</td><td> 47566051</td><td> 4</td><td> 2</td><td>23457</td></tr>\n",
       "\t<tr><td>Zm00001d028818</td><td>1 </td><td> 47566046</td><td> 47589502</td><td>Row-03</td><td> 47566043</td><td> 47566051</td><td> 7</td><td> 4</td><td>23457</td></tr>\n",
       "\t<tr><td>Zm00001d028962</td><td>1 </td><td> 53030837</td><td> 53032126</td><td>Col-03</td><td> 53032126</td><td> 53032134</td><td>14</td><td>17</td><td> 1290</td></tr>\n",
       "\t<tr><td>Zm00001d028962</td><td>1 </td><td> 53030837</td><td> 53032126</td><td>Row-03</td><td> 53032126</td><td> 53032134</td><td>35</td><td>27</td><td> 1290</td></tr>\n",
       "\t<tr><td>Zm00001d029144</td><td>1 </td><td> 60427117</td><td> 60431202</td><td>Col-02</td><td> 60431195</td><td> 60431203</td><td>12</td><td> 6</td><td> 4086</td></tr>\n",
       "\t<tr><td>Zm00001d029144</td><td>1 </td><td> 60427117</td><td> 60431202</td><td>Row-03</td><td> 60431195</td><td> 60431203</td><td> 6</td><td> 2</td><td> 4086</td></tr>\n",
       "\t<tr><td>Zm00001d022970</td><td>1 </td><td>261786400</td><td>261787460</td><td>Col-02</td><td>261786393</td><td>261786401</td><td> 3</td><td> 2</td><td> 1061</td></tr>\n",
       "\t<tr><td>Zm00001d022970</td><td>1 </td><td>261786400</td><td>261787460</td><td>Row-03</td><td>261786393</td><td>261786401</td><td> 4</td><td> 2</td><td> 1061</td></tr>\n",
       "\t<tr><td>Zm00001d033420</td><td>1 </td><td>262569031</td><td>262571498</td><td>Col-02</td><td>262569026</td><td>262569034</td><td> 8</td><td> 2</td><td> 2468</td></tr>\n",
       "\t<tr><td>Zm00001d033420</td><td>1 </td><td>262569031</td><td>262571498</td><td>Row-02</td><td>262569026</td><td>262569034</td><td>23</td><td>13</td><td> 2468</td></tr>\n",
       "\t<tr><td>Zm00001d034432</td><td>1 </td><td>292893097</td><td>292904934</td><td>Col-01</td><td>292904929</td><td>292904937</td><td> 2</td><td> 2</td><td>11838</td></tr>\n",
       "\t<tr><td>Zm00001d034432</td><td>1 </td><td>292893097</td><td>292904934</td><td>Row-01</td><td>292904929</td><td>292904937</td><td> 4</td><td> 3</td><td>11838</td></tr>\n",
       "\t<tr><td>Zm00001d034638</td><td>1 </td><td>298677159</td><td>298680740</td><td>Col-03</td><td>298677155</td><td>298677163</td><td> 3</td><td> 3</td><td> 3582</td></tr>\n",
       "\t<tr><td>Zm00001d034638</td><td>1 </td><td>298677159</td><td>298680740</td><td>Row-02</td><td>298677155</td><td>298677163</td><td>77</td><td>18</td><td> 3582</td></tr>\n",
       "\t<tr><td>Zm00001d024593</td><td>10</td><td> 79588608</td><td> 79589422</td><td>Col-01</td><td> 79589421</td><td> 79589429</td><td> 5</td><td>14</td><td>  815</td></tr>\n",
       "\t<tr><td>Zm00001d024593</td><td>10</td><td> 79588608</td><td> 79589422</td><td>Row-03</td><td> 79589421</td><td> 79589429</td><td> 4</td><td>13</td><td>  815</td></tr>\n",
       "\t<tr><td>Zm00001d024676</td><td>10</td><td> 83290724</td><td> 83291829</td><td>Col-03</td><td> 83290721</td><td> 83290729</td><td> 6</td><td> 9</td><td> 1106</td></tr>\n",
       "\t<tr><td>Zm00001d024676</td><td>10</td><td> 83290724</td><td> 83291829</td><td>Row-01</td><td> 83290721</td><td> 83290729</td><td> 8</td><td>12</td><td> 1106</td></tr>\n",
       "\t<tr><td>Zm00001d024933</td><td>10</td><td> 95587713</td><td> 95593434</td><td>Col-03</td><td> 95587708</td><td> 95587716</td><td>35</td><td>21</td><td> 5722</td></tr>\n",
       "\t<tr><td>Zm00001d024933</td><td>10</td><td> 95587713</td><td> 95593434</td><td>Row-01</td><td> 95587708</td><td> 95587716</td><td> 3</td><td> 2</td><td> 5722</td></tr>\n",
       "\t<tr><td>Zm00001d004137</td><td>2 </td><td> 85731223</td><td> 85732587</td><td>Col-01</td><td> 85731220</td><td> 85731228</td><td> 6</td><td> 9</td><td> 1365</td></tr>\n",
       "\t<tr><td>Zm00001d004137</td><td>2 </td><td> 85731223</td><td> 85732587</td><td>Row-01</td><td> 85731220</td><td> 85731228</td><td> 4</td><td> 4</td><td> 1365</td></tr>\n",
       "\t<tr><td>Zm00001d004812</td><td>2 </td><td>140809252</td><td>140813969</td><td>Col-03</td><td>140809245</td><td>140809253</td><td>15</td><td>12</td><td> 4718</td></tr>\n",
       "\t<tr><td>Zm00001d004812</td><td>2 </td><td>140809252</td><td>140813969</td><td>Row-03</td><td>140809245</td><td>140809253</td><td>30</td><td> 2</td><td> 4718</td></tr>\n",
       "\t<tr><td>Zm00001d006249</td><td>2 </td><td>202837832</td><td>202840525</td><td>Col-02</td><td>202840523</td><td>202840531</td><td> 4</td><td> 2</td><td> 2694</td></tr>\n",
       "\t<tr><td>Zm00001d006249</td><td>2 </td><td>202837832</td><td>202840525</td><td>Row-02</td><td>202840523</td><td>202840531</td><td>14</td><td>16</td><td> 2694</td></tr>\n",
       "\t<tr><td>Zm00001d041829</td><td>3 </td><td>139490055</td><td>139496394</td><td>Col-04</td><td>139496394</td><td>139496402</td><td> 4</td><td> 6</td><td> 6340</td></tr>\n",
       "\t<tr><td>Zm00001d041829</td><td>3 </td><td>139490055</td><td>139496394</td><td>Row-01</td><td>139496394</td><td>139496402</td><td> 2</td><td> 3</td><td> 6340</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Zm00001d054070</td><td>4          </td><td>245864833</td><td>245876851</td><td>Col-01</td><td>245864832</td><td>245864840</td><td> 4</td><td> 7</td><td>12019</td></tr>\n",
       "\t<tr><td>Zm00001d054070</td><td>4          </td><td>245864833</td><td>245876851</td><td>Row-03</td><td>245864832</td><td>245864840</td><td> 2</td><td> 5</td><td>12019</td></tr>\n",
       "\t<tr><td>Zm00001d013646</td><td>5          </td><td> 16124142</td><td> 16125803</td><td>Col-03</td><td> 16124139</td><td> 16124147</td><td>73</td><td>77</td><td> 1662</td></tr>\n",
       "\t<tr><td>Zm00001d013646</td><td>5          </td><td> 16124142</td><td> 16125803</td><td>Row-01</td><td> 16124139</td><td> 16124147</td><td>75</td><td>61</td><td> 1662</td></tr>\n",
       "\t<tr><td>Zm00001d017021</td><td>5          </td><td>182894380</td><td>182895386</td><td>Col-01</td><td>182894374</td><td>182894382</td><td>27</td><td>19</td><td> 1007</td></tr>\n",
       "\t<tr><td>Zm00001d017021</td><td>5          </td><td>182894380</td><td>182895386</td><td>Row-04</td><td>182894374</td><td>182894382</td><td> 2</td><td> 4</td><td> 1007</td></tr>\n",
       "\t<tr><td>Zm00001d018207</td><td>5          </td><td>216597825</td><td>216606598</td><td>Col-03</td><td>216606593</td><td>216606601</td><td> 2</td><td> 2</td><td> 8774</td></tr>\n",
       "\t<tr><td>Zm00001d018207</td><td>5          </td><td>216597825</td><td>216606598</td><td>Row-01</td><td>216606593</td><td>216606601</td><td> 3</td><td> 2</td><td> 8774</td></tr>\n",
       "\t<tr><td>Zm00001d018420</td><td>5          </td><td>220379839</td><td>220387369</td><td>Col-01</td><td>220387364</td><td>220387372</td><td>22</td><td>12</td><td> 7531</td></tr>\n",
       "\t<tr><td>Zm00001d018420</td><td>5          </td><td>220379839</td><td>220387369</td><td>Row-01</td><td>220387364</td><td>220387372</td><td>23</td><td>31</td><td> 7531</td></tr>\n",
       "\t<tr><td>Zm00001d038189</td><td>6          </td><td>150913450</td><td>150943935</td><td>Col-03</td><td>150943934</td><td>150943942</td><td>13</td><td> 7</td><td>30486</td></tr>\n",
       "\t<tr><td>Zm00001d038189</td><td>6          </td><td>150913450</td><td>150943935</td><td>Row-03</td><td>150943934</td><td>150943942</td><td> 2</td><td> 2</td><td>30486</td></tr>\n",
       "\t<tr><td>Zm00001d038732</td><td>6          </td><td>163536956</td><td>163538558</td><td>Col-02</td><td>163538551</td><td>163538559</td><td>15</td><td> 4</td><td> 1603</td></tr>\n",
       "\t<tr><td>Zm00001d038732</td><td>6          </td><td>163536956</td><td>163538558</td><td>Row-02</td><td>163538551</td><td>163538559</td><td>16</td><td>20</td><td> 1603</td></tr>\n",
       "\t<tr><td>Zm00001d039020</td><td>6          </td><td>168822275</td><td>168828891</td><td>Col-04</td><td>168822272</td><td>168822280</td><td> 4</td><td> 4</td><td> 6617</td></tr>\n",
       "\t<tr><td>Zm00001d039020</td><td>6          </td><td>168822275</td><td>168828891</td><td>Row-01</td><td>168822272</td><td>168822280</td><td> 2</td><td> 2</td><td> 6617</td></tr>\n",
       "\t<tr><td>Zm00001d018902</td><td>7          </td><td>  8758340</td><td>  8762850</td><td>Col-02</td><td>  8762848</td><td>  8762856</td><td>11</td><td>19</td><td> 4511</td></tr>\n",
       "\t<tr><td>Zm00001d018902</td><td>7          </td><td>  8758340</td><td>  8762850</td><td>Row-03</td><td>  8762848</td><td>  8762856</td><td> 7</td><td> 8</td><td> 4511</td></tr>\n",
       "\t<tr><td>Zm00001d019644</td><td>7          </td><td> 47868260</td><td> 47872065</td><td>Col-02</td><td> 47868258</td><td> 47868266</td><td> 2</td><td> 4</td><td> 3806</td></tr>\n",
       "\t<tr><td>Zm00001d019644</td><td>7          </td><td> 47868260</td><td> 47872065</td><td>Row-04</td><td> 47868258</td><td> 47868266</td><td> 2</td><td> 3</td><td> 3806</td></tr>\n",
       "\t<tr><td>Zm00001d019719</td><td>7          </td><td> 53590581</td><td> 53601030</td><td>Col-04</td><td> 53601027</td><td> 53601035</td><td>57</td><td>44</td><td>10450</td></tr>\n",
       "\t<tr><td>Zm00001d019719</td><td>7          </td><td> 53590581</td><td> 53601030</td><td>Row-03</td><td> 53601027</td><td> 53601035</td><td> 9</td><td> 5</td><td>10450</td></tr>\n",
       "\t<tr><td>Zm00001d009413</td><td>8          </td><td> 63468493</td><td> 63474399</td><td>Col-03</td><td> 63468486</td><td> 63468494</td><td> 7</td><td>12</td><td> 5907</td></tr>\n",
       "\t<tr><td>Zm00001d009413</td><td>8          </td><td> 63468493</td><td> 63474399</td><td>Row-01</td><td> 63468486</td><td> 63468494</td><td> 2</td><td> 3</td><td> 5907</td></tr>\n",
       "\t<tr><td>Zm00001d044829</td><td>9          </td><td>  4331182</td><td>  4332060</td><td>Col-04</td><td>  4331175</td><td>  4331183</td><td>56</td><td>19</td><td>  879</td></tr>\n",
       "\t<tr><td>Zm00001d044829</td><td>9          </td><td>  4331182</td><td>  4332060</td><td>Row-01</td><td>  4331175</td><td>  4331183</td><td>10</td><td> 6</td><td>  879</td></tr>\n",
       "\t<tr><td>Zm00001d046590</td><td>9          </td><td> 97452571</td><td> 97463456</td><td>Col-02</td><td> 97463450</td><td> 97463458</td><td> 2</td><td> 2</td><td>10886</td></tr>\n",
       "\t<tr><td>Zm00001d046590</td><td>9          </td><td> 97452571</td><td> 97463456</td><td>Row-01</td><td> 97463450</td><td> 97463458</td><td> 2</td><td> 2</td><td>10886</td></tr>\n",
       "\t<tr><td>Zm00001d000150</td><td>B73V4_ctg75</td><td>   116649</td><td>   127114</td><td>Col-04</td><td>   127109</td><td>   127117</td><td> 3</td><td> 2</td><td>10466</td></tr>\n",
       "\t<tr><td>Zm00001d000150</td><td>B73V4_ctg75</td><td>   116649</td><td>   127114</td><td>Row-04</td><td>   127109</td><td>   127117</td><td> 3</td><td> 2</td><td>10466</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 80 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " GeneID & Chr & Start & End & Sample & InsertionStart & InsertionEnd & StartReads & EndReads & Gene\\_length\\\\\n",
       " <chr> & <chr> & <int> & <int> & <chr> & <int> & <int> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Zm00001d027801 & 1  &  13815367 &  13827552 & Col-02 &  13815363 &  13815371 &  6 &  9 & 12186\\\\\n",
       "\t Zm00001d027801 & 1  &  13815367 &  13827552 & Row-01 &  13815363 &  13815371 &  6 &  4 & 12186\\\\\n",
       "\t Zm00001d028818 & 1  &  47566046 &  47589502 & Col-01 &  47566043 &  47566051 &  4 &  2 & 23457\\\\\n",
       "\t Zm00001d028818 & 1  &  47566046 &  47589502 & Row-03 &  47566043 &  47566051 &  7 &  4 & 23457\\\\\n",
       "\t Zm00001d028962 & 1  &  53030837 &  53032126 & Col-03 &  53032126 &  53032134 & 14 & 17 &  1290\\\\\n",
       "\t Zm00001d028962 & 1  &  53030837 &  53032126 & Row-03 &  53032126 &  53032134 & 35 & 27 &  1290\\\\\n",
       "\t Zm00001d029144 & 1  &  60427117 &  60431202 & Col-02 &  60431195 &  60431203 & 12 &  6 &  4086\\\\\n",
       "\t Zm00001d029144 & 1  &  60427117 &  60431202 & Row-03 &  60431195 &  60431203 &  6 &  2 &  4086\\\\\n",
       "\t Zm00001d022970 & 1  & 261786400 & 261787460 & Col-02 & 261786393 & 261786401 &  3 &  2 &  1061\\\\\n",
       "\t Zm00001d022970 & 1  & 261786400 & 261787460 & Row-03 & 261786393 & 261786401 &  4 &  2 &  1061\\\\\n",
       "\t Zm00001d033420 & 1  & 262569031 & 262571498 & Col-02 & 262569026 & 262569034 &  8 &  2 &  2468\\\\\n",
       "\t Zm00001d033420 & 1  & 262569031 & 262571498 & Row-02 & 262569026 & 262569034 & 23 & 13 &  2468\\\\\n",
       "\t Zm00001d034432 & 1  & 292893097 & 292904934 & Col-01 & 292904929 & 292904937 &  2 &  2 & 11838\\\\\n",
       "\t Zm00001d034432 & 1  & 292893097 & 292904934 & Row-01 & 292904929 & 292904937 &  4 &  3 & 11838\\\\\n",
       "\t Zm00001d034638 & 1  & 298677159 & 298680740 & Col-03 & 298677155 & 298677163 &  3 &  3 &  3582\\\\\n",
       "\t Zm00001d034638 & 1  & 298677159 & 298680740 & Row-02 & 298677155 & 298677163 & 77 & 18 &  3582\\\\\n",
       "\t Zm00001d024593 & 10 &  79588608 &  79589422 & Col-01 &  79589421 &  79589429 &  5 & 14 &   815\\\\\n",
       "\t Zm00001d024593 & 10 &  79588608 &  79589422 & Row-03 &  79589421 &  79589429 &  4 & 13 &   815\\\\\n",
       "\t Zm00001d024676 & 10 &  83290724 &  83291829 & Col-03 &  83290721 &  83290729 &  6 &  9 &  1106\\\\\n",
       "\t Zm00001d024676 & 10 &  83290724 &  83291829 & Row-01 &  83290721 &  83290729 &  8 & 12 &  1106\\\\\n",
       "\t Zm00001d024933 & 10 &  95587713 &  95593434 & Col-03 &  95587708 &  95587716 & 35 & 21 &  5722\\\\\n",
       "\t Zm00001d024933 & 10 &  95587713 &  95593434 & Row-01 &  95587708 &  95587716 &  3 &  2 &  5722\\\\\n",
       "\t Zm00001d004137 & 2  &  85731223 &  85732587 & Col-01 &  85731220 &  85731228 &  6 &  9 &  1365\\\\\n",
       "\t Zm00001d004137 & 2  &  85731223 &  85732587 & Row-01 &  85731220 &  85731228 &  4 &  4 &  1365\\\\\n",
       "\t Zm00001d004812 & 2  & 140809252 & 140813969 & Col-03 & 140809245 & 140809253 & 15 & 12 &  4718\\\\\n",
       "\t Zm00001d004812 & 2  & 140809252 & 140813969 & Row-03 & 140809245 & 140809253 & 30 &  2 &  4718\\\\\n",
       "\t Zm00001d006249 & 2  & 202837832 & 202840525 & Col-02 & 202840523 & 202840531 &  4 &  2 &  2694\\\\\n",
       "\t Zm00001d006249 & 2  & 202837832 & 202840525 & Row-02 & 202840523 & 202840531 & 14 & 16 &  2694\\\\\n",
       "\t Zm00001d041829 & 3  & 139490055 & 139496394 & Col-04 & 139496394 & 139496402 &  4 &  6 &  6340\\\\\n",
       "\t Zm00001d041829 & 3  & 139490055 & 139496394 & Row-01 & 139496394 & 139496402 &  2 &  3 &  6340\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Zm00001d054070 & 4           & 245864833 & 245876851 & Col-01 & 245864832 & 245864840 &  4 &  7 & 12019\\\\\n",
       "\t Zm00001d054070 & 4           & 245864833 & 245876851 & Row-03 & 245864832 & 245864840 &  2 &  5 & 12019\\\\\n",
       "\t Zm00001d013646 & 5           &  16124142 &  16125803 & Col-03 &  16124139 &  16124147 & 73 & 77 &  1662\\\\\n",
       "\t Zm00001d013646 & 5           &  16124142 &  16125803 & Row-01 &  16124139 &  16124147 & 75 & 61 &  1662\\\\\n",
       "\t Zm00001d017021 & 5           & 182894380 & 182895386 & Col-01 & 182894374 & 182894382 & 27 & 19 &  1007\\\\\n",
       "\t Zm00001d017021 & 5           & 182894380 & 182895386 & Row-04 & 182894374 & 182894382 &  2 &  4 &  1007\\\\\n",
       "\t Zm00001d018207 & 5           & 216597825 & 216606598 & Col-03 & 216606593 & 216606601 &  2 &  2 &  8774\\\\\n",
       "\t Zm00001d018207 & 5           & 216597825 & 216606598 & Row-01 & 216606593 & 216606601 &  3 &  2 &  8774\\\\\n",
       "\t Zm00001d018420 & 5           & 220379839 & 220387369 & Col-01 & 220387364 & 220387372 & 22 & 12 &  7531\\\\\n",
       "\t Zm00001d018420 & 5           & 220379839 & 220387369 & Row-01 & 220387364 & 220387372 & 23 & 31 &  7531\\\\\n",
       "\t Zm00001d038189 & 6           & 150913450 & 150943935 & Col-03 & 150943934 & 150943942 & 13 &  7 & 30486\\\\\n",
       "\t Zm00001d038189 & 6           & 150913450 & 150943935 & Row-03 & 150943934 & 150943942 &  2 &  2 & 30486\\\\\n",
       "\t Zm00001d038732 & 6           & 163536956 & 163538558 & Col-02 & 163538551 & 163538559 & 15 &  4 &  1603\\\\\n",
       "\t Zm00001d038732 & 6           & 163536956 & 163538558 & Row-02 & 163538551 & 163538559 & 16 & 20 &  1603\\\\\n",
       "\t Zm00001d039020 & 6           & 168822275 & 168828891 & Col-04 & 168822272 & 168822280 &  4 &  4 &  6617\\\\\n",
       "\t Zm00001d039020 & 6           & 168822275 & 168828891 & Row-01 & 168822272 & 168822280 &  2 &  2 &  6617\\\\\n",
       "\t Zm00001d018902 & 7           &   8758340 &   8762850 & Col-02 &   8762848 &   8762856 & 11 & 19 &  4511\\\\\n",
       "\t Zm00001d018902 & 7           &   8758340 &   8762850 & Row-03 &   8762848 &   8762856 &  7 &  8 &  4511\\\\\n",
       "\t Zm00001d019644 & 7           &  47868260 &  47872065 & Col-02 &  47868258 &  47868266 &  2 &  4 &  3806\\\\\n",
       "\t Zm00001d019644 & 7           &  47868260 &  47872065 & Row-04 &  47868258 &  47868266 &  2 &  3 &  3806\\\\\n",
       "\t Zm00001d019719 & 7           &  53590581 &  53601030 & Col-04 &  53601027 &  53601035 & 57 & 44 & 10450\\\\\n",
       "\t Zm00001d019719 & 7           &  53590581 &  53601030 & Row-03 &  53601027 &  53601035 &  9 &  5 & 10450\\\\\n",
       "\t Zm00001d009413 & 8           &  63468493 &  63474399 & Col-03 &  63468486 &  63468494 &  7 & 12 &  5907\\\\\n",
       "\t Zm00001d009413 & 8           &  63468493 &  63474399 & Row-01 &  63468486 &  63468494 &  2 &  3 &  5907\\\\\n",
       "\t Zm00001d044829 & 9           &   4331182 &   4332060 & Col-04 &   4331175 &   4331183 & 56 & 19 &   879\\\\\n",
       "\t Zm00001d044829 & 9           &   4331182 &   4332060 & Row-01 &   4331175 &   4331183 & 10 &  6 &   879\\\\\n",
       "\t Zm00001d046590 & 9           &  97452571 &  97463456 & Col-02 &  97463450 &  97463458 &  2 &  2 & 10886\\\\\n",
       "\t Zm00001d046590 & 9           &  97452571 &  97463456 & Row-01 &  97463450 &  97463458 &  2 &  2 & 10886\\\\\n",
       "\t Zm00001d000150 & B73V4\\_ctg75 &    116649 &    127114 & Col-04 &    127109 &    127117 &  3 &  2 & 10466\\\\\n",
       "\t Zm00001d000150 & B73V4\\_ctg75 &    116649 &    127114 & Row-04 &    127109 &    127117 &  3 &  2 & 10466\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 80 × 10\n",
       "\n",
       "| GeneID &lt;chr&gt; | Chr &lt;chr&gt; | Start &lt;int&gt; | End &lt;int&gt; | Sample &lt;chr&gt; | InsertionStart &lt;int&gt; | InsertionEnd &lt;int&gt; | StartReads &lt;int&gt; | EndReads &lt;int&gt; | Gene_length &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| Zm00001d027801 | 1  |  13815367 |  13827552 | Col-02 |  13815363 |  13815371 |  6 |  9 | 12186 |\n",
       "| Zm00001d027801 | 1  |  13815367 |  13827552 | Row-01 |  13815363 |  13815371 |  6 |  4 | 12186 |\n",
       "| Zm00001d028818 | 1  |  47566046 |  47589502 | Col-01 |  47566043 |  47566051 |  4 |  2 | 23457 |\n",
       "| Zm00001d028818 | 1  |  47566046 |  47589502 | Row-03 |  47566043 |  47566051 |  7 |  4 | 23457 |\n",
       "| Zm00001d028962 | 1  |  53030837 |  53032126 | Col-03 |  53032126 |  53032134 | 14 | 17 |  1290 |\n",
       "| Zm00001d028962 | 1  |  53030837 |  53032126 | Row-03 |  53032126 |  53032134 | 35 | 27 |  1290 |\n",
       "| Zm00001d029144 | 1  |  60427117 |  60431202 | Col-02 |  60431195 |  60431203 | 12 |  6 |  4086 |\n",
       "| Zm00001d029144 | 1  |  60427117 |  60431202 | Row-03 |  60431195 |  60431203 |  6 |  2 |  4086 |\n",
       "| Zm00001d022970 | 1  | 261786400 | 261787460 | Col-02 | 261786393 | 261786401 |  3 |  2 |  1061 |\n",
       "| Zm00001d022970 | 1  | 261786400 | 261787460 | Row-03 | 261786393 | 261786401 |  4 |  2 |  1061 |\n",
       "| Zm00001d033420 | 1  | 262569031 | 262571498 | Col-02 | 262569026 | 262569034 |  8 |  2 |  2468 |\n",
       "| Zm00001d033420 | 1  | 262569031 | 262571498 | Row-02 | 262569026 | 262569034 | 23 | 13 |  2468 |\n",
       "| Zm00001d034432 | 1  | 292893097 | 292904934 | Col-01 | 292904929 | 292904937 |  2 |  2 | 11838 |\n",
       "| Zm00001d034432 | 1  | 292893097 | 292904934 | Row-01 | 292904929 | 292904937 |  4 |  3 | 11838 |\n",
       "| Zm00001d034638 | 1  | 298677159 | 298680740 | Col-03 | 298677155 | 298677163 |  3 |  3 |  3582 |\n",
       "| Zm00001d034638 | 1  | 298677159 | 298680740 | Row-02 | 298677155 | 298677163 | 77 | 18 |  3582 |\n",
       "| Zm00001d024593 | 10 |  79588608 |  79589422 | Col-01 |  79589421 |  79589429 |  5 | 14 |   815 |\n",
       "| Zm00001d024593 | 10 |  79588608 |  79589422 | Row-03 |  79589421 |  79589429 |  4 | 13 |   815 |\n",
       "| Zm00001d024676 | 10 |  83290724 |  83291829 | Col-03 |  83290721 |  83290729 |  6 |  9 |  1106 |\n",
       "| Zm00001d024676 | 10 |  83290724 |  83291829 | Row-01 |  83290721 |  83290729 |  8 | 12 |  1106 |\n",
       "| Zm00001d024933 | 10 |  95587713 |  95593434 | Col-03 |  95587708 |  95587716 | 35 | 21 |  5722 |\n",
       "| Zm00001d024933 | 10 |  95587713 |  95593434 | Row-01 |  95587708 |  95587716 |  3 |  2 |  5722 |\n",
       "| Zm00001d004137 | 2  |  85731223 |  85732587 | Col-01 |  85731220 |  85731228 |  6 |  9 |  1365 |\n",
       "| Zm00001d004137 | 2  |  85731223 |  85732587 | Row-01 |  85731220 |  85731228 |  4 |  4 |  1365 |\n",
       "| Zm00001d004812 | 2  | 140809252 | 140813969 | Col-03 | 140809245 | 140809253 | 15 | 12 |  4718 |\n",
       "| Zm00001d004812 | 2  | 140809252 | 140813969 | Row-03 | 140809245 | 140809253 | 30 |  2 |  4718 |\n",
       "| Zm00001d006249 | 2  | 202837832 | 202840525 | Col-02 | 202840523 | 202840531 |  4 |  2 |  2694 |\n",
       "| Zm00001d006249 | 2  | 202837832 | 202840525 | Row-02 | 202840523 | 202840531 | 14 | 16 |  2694 |\n",
       "| Zm00001d041829 | 3  | 139490055 | 139496394 | Col-04 | 139496394 | 139496402 |  4 |  6 |  6340 |\n",
       "| Zm00001d041829 | 3  | 139490055 | 139496394 | Row-01 | 139496394 | 139496402 |  2 |  3 |  6340 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Zm00001d054070 | 4           | 245864833 | 245876851 | Col-01 | 245864832 | 245864840 |  4 |  7 | 12019 |\n",
       "| Zm00001d054070 | 4           | 245864833 | 245876851 | Row-03 | 245864832 | 245864840 |  2 |  5 | 12019 |\n",
       "| Zm00001d013646 | 5           |  16124142 |  16125803 | Col-03 |  16124139 |  16124147 | 73 | 77 |  1662 |\n",
       "| Zm00001d013646 | 5           |  16124142 |  16125803 | Row-01 |  16124139 |  16124147 | 75 | 61 |  1662 |\n",
       "| Zm00001d017021 | 5           | 182894380 | 182895386 | Col-01 | 182894374 | 182894382 | 27 | 19 |  1007 |\n",
       "| Zm00001d017021 | 5           | 182894380 | 182895386 | Row-04 | 182894374 | 182894382 |  2 |  4 |  1007 |\n",
       "| Zm00001d018207 | 5           | 216597825 | 216606598 | Col-03 | 216606593 | 216606601 |  2 |  2 |  8774 |\n",
       "| Zm00001d018207 | 5           | 216597825 | 216606598 | Row-01 | 216606593 | 216606601 |  3 |  2 |  8774 |\n",
       "| Zm00001d018420 | 5           | 220379839 | 220387369 | Col-01 | 220387364 | 220387372 | 22 | 12 |  7531 |\n",
       "| Zm00001d018420 | 5           | 220379839 | 220387369 | Row-01 | 220387364 | 220387372 | 23 | 31 |  7531 |\n",
       "| Zm00001d038189 | 6           | 150913450 | 150943935 | Col-03 | 150943934 | 150943942 | 13 |  7 | 30486 |\n",
       "| Zm00001d038189 | 6           | 150913450 | 150943935 | Row-03 | 150943934 | 150943942 |  2 |  2 | 30486 |\n",
       "| Zm00001d038732 | 6           | 163536956 | 163538558 | Col-02 | 163538551 | 163538559 | 15 |  4 |  1603 |\n",
       "| Zm00001d038732 | 6           | 163536956 | 163538558 | Row-02 | 163538551 | 163538559 | 16 | 20 |  1603 |\n",
       "| Zm00001d039020 | 6           | 168822275 | 168828891 | Col-04 | 168822272 | 168822280 |  4 |  4 |  6617 |\n",
       "| Zm00001d039020 | 6           | 168822275 | 168828891 | Row-01 | 168822272 | 168822280 |  2 |  2 |  6617 |\n",
       "| Zm00001d018902 | 7           |   8758340 |   8762850 | Col-02 |   8762848 |   8762856 | 11 | 19 |  4511 |\n",
       "| Zm00001d018902 | 7           |   8758340 |   8762850 | Row-03 |   8762848 |   8762856 |  7 |  8 |  4511 |\n",
       "| Zm00001d019644 | 7           |  47868260 |  47872065 | Col-02 |  47868258 |  47868266 |  2 |  4 |  3806 |\n",
       "| Zm00001d019644 | 7           |  47868260 |  47872065 | Row-04 |  47868258 |  47868266 |  2 |  3 |  3806 |\n",
       "| Zm00001d019719 | 7           |  53590581 |  53601030 | Col-04 |  53601027 |  53601035 | 57 | 44 | 10450 |\n",
       "| Zm00001d019719 | 7           |  53590581 |  53601030 | Row-03 |  53601027 |  53601035 |  9 |  5 | 10450 |\n",
       "| Zm00001d009413 | 8           |  63468493 |  63474399 | Col-03 |  63468486 |  63468494 |  7 | 12 |  5907 |\n",
       "| Zm00001d009413 | 8           |  63468493 |  63474399 | Row-01 |  63468486 |  63468494 |  2 |  3 |  5907 |\n",
       "| Zm00001d044829 | 9           |   4331182 |   4332060 | Col-04 |   4331175 |   4331183 | 56 | 19 |   879 |\n",
       "| Zm00001d044829 | 9           |   4331182 |   4332060 | Row-01 |   4331175 |   4331183 | 10 |  6 |   879 |\n",
       "| Zm00001d046590 | 9           |  97452571 |  97463456 | Col-02 |  97463450 |  97463458 |  2 |  2 | 10886 |\n",
       "| Zm00001d046590 | 9           |  97452571 |  97463456 | Row-01 |  97463450 |  97463458 |  2 |  2 | 10886 |\n",
       "| Zm00001d000150 | B73V4_ctg75 |    116649 |    127114 | Col-04 |    127109 |    127117 |  3 |  2 | 10466 |\n",
       "| Zm00001d000150 | B73V4_ctg75 |    116649 |    127114 | Row-04 |    127109 |    127117 |  3 |  2 | 10466 |\n",
       "\n"
      ],
      "text/plain": [
       "   GeneID         Chr         Start     End       Sample InsertionStart\n",
       "1  Zm00001d027801 1            13815367  13827552 Col-02  13815363     \n",
       "2  Zm00001d027801 1            13815367  13827552 Row-01  13815363     \n",
       "3  Zm00001d028818 1            47566046  47589502 Col-01  47566043     \n",
       "4  Zm00001d028818 1            47566046  47589502 Row-03  47566043     \n",
       "5  Zm00001d028962 1            53030837  53032126 Col-03  53032126     \n",
       "6  Zm00001d028962 1            53030837  53032126 Row-03  53032126     \n",
       "7  Zm00001d029144 1            60427117  60431202 Col-02  60431195     \n",
       "8  Zm00001d029144 1            60427117  60431202 Row-03  60431195     \n",
       "9  Zm00001d022970 1           261786400 261787460 Col-02 261786393     \n",
       "10 Zm00001d022970 1           261786400 261787460 Row-03 261786393     \n",
       "11 Zm00001d033420 1           262569031 262571498 Col-02 262569026     \n",
       "12 Zm00001d033420 1           262569031 262571498 Row-02 262569026     \n",
       "13 Zm00001d034432 1           292893097 292904934 Col-01 292904929     \n",
       "14 Zm00001d034432 1           292893097 292904934 Row-01 292904929     \n",
       "15 Zm00001d034638 1           298677159 298680740 Col-03 298677155     \n",
       "16 Zm00001d034638 1           298677159 298680740 Row-02 298677155     \n",
       "17 Zm00001d024593 10           79588608  79589422 Col-01  79589421     \n",
       "18 Zm00001d024593 10           79588608  79589422 Row-03  79589421     \n",
       "19 Zm00001d024676 10           83290724  83291829 Col-03  83290721     \n",
       "20 Zm00001d024676 10           83290724  83291829 Row-01  83290721     \n",
       "21 Zm00001d024933 10           95587713  95593434 Col-03  95587708     \n",
       "22 Zm00001d024933 10           95587713  95593434 Row-01  95587708     \n",
       "23 Zm00001d004137 2            85731223  85732587 Col-01  85731220     \n",
       "24 Zm00001d004137 2            85731223  85732587 Row-01  85731220     \n",
       "25 Zm00001d004812 2           140809252 140813969 Col-03 140809245     \n",
       "26 Zm00001d004812 2           140809252 140813969 Row-03 140809245     \n",
       "27 Zm00001d006249 2           202837832 202840525 Col-02 202840523     \n",
       "28 Zm00001d006249 2           202837832 202840525 Row-02 202840523     \n",
       "29 Zm00001d041829 3           139490055 139496394 Col-04 139496394     \n",
       "30 Zm00001d041829 3           139490055 139496394 Row-01 139496394     \n",
       "⋮  ⋮              ⋮           ⋮         ⋮         ⋮      ⋮             \n",
       "51 Zm00001d054070 4           245864833 245876851 Col-01 245864832     \n",
       "52 Zm00001d054070 4           245864833 245876851 Row-03 245864832     \n",
       "53 Zm00001d013646 5            16124142  16125803 Col-03  16124139     \n",
       "54 Zm00001d013646 5            16124142  16125803 Row-01  16124139     \n",
       "55 Zm00001d017021 5           182894380 182895386 Col-01 182894374     \n",
       "56 Zm00001d017021 5           182894380 182895386 Row-04 182894374     \n",
       "57 Zm00001d018207 5           216597825 216606598 Col-03 216606593     \n",
       "58 Zm00001d018207 5           216597825 216606598 Row-01 216606593     \n",
       "59 Zm00001d018420 5           220379839 220387369 Col-01 220387364     \n",
       "60 Zm00001d018420 5           220379839 220387369 Row-01 220387364     \n",
       "61 Zm00001d038189 6           150913450 150943935 Col-03 150943934     \n",
       "62 Zm00001d038189 6           150913450 150943935 Row-03 150943934     \n",
       "63 Zm00001d038732 6           163536956 163538558 Col-02 163538551     \n",
       "64 Zm00001d038732 6           163536956 163538558 Row-02 163538551     \n",
       "65 Zm00001d039020 6           168822275 168828891 Col-04 168822272     \n",
       "66 Zm00001d039020 6           168822275 168828891 Row-01 168822272     \n",
       "67 Zm00001d018902 7             8758340   8762850 Col-02   8762848     \n",
       "68 Zm00001d018902 7             8758340   8762850 Row-03   8762848     \n",
       "69 Zm00001d019644 7            47868260  47872065 Col-02  47868258     \n",
       "70 Zm00001d019644 7            47868260  47872065 Row-04  47868258     \n",
       "71 Zm00001d019719 7            53590581  53601030 Col-04  53601027     \n",
       "72 Zm00001d019719 7            53590581  53601030 Row-03  53601027     \n",
       "73 Zm00001d009413 8            63468493  63474399 Col-03  63468486     \n",
       "74 Zm00001d009413 8            63468493  63474399 Row-01  63468486     \n",
       "75 Zm00001d044829 9             4331182   4332060 Col-04   4331175     \n",
       "76 Zm00001d044829 9             4331182   4332060 Row-01   4331175     \n",
       "77 Zm00001d046590 9            97452571  97463456 Col-02  97463450     \n",
       "78 Zm00001d046590 9            97452571  97463456 Row-01  97463450     \n",
       "79 Zm00001d000150 B73V4_ctg75    116649    127114 Col-04    127109     \n",
       "80 Zm00001d000150 B73V4_ctg75    116649    127114 Row-04    127109     \n",
       "   InsertionEnd StartReads EndReads Gene_length\n",
       "1   13815371     6          9       12186      \n",
       "2   13815371     6          4       12186      \n",
       "3   47566051     4          2       23457      \n",
       "4   47566051     7          4       23457      \n",
       "5   53032134    14         17        1290      \n",
       "6   53032134    35         27        1290      \n",
       "7   60431203    12          6        4086      \n",
       "8   60431203     6          2        4086      \n",
       "9  261786401     3          2        1061      \n",
       "10 261786401     4          2        1061      \n",
       "11 262569034     8          2        2468      \n",
       "12 262569034    23         13        2468      \n",
       "13 292904937     2          2       11838      \n",
       "14 292904937     4          3       11838      \n",
       "15 298677163     3          3        3582      \n",
       "16 298677163    77         18        3582      \n",
       "17  79589429     5         14         815      \n",
       "18  79589429     4         13         815      \n",
       "19  83290729     6          9        1106      \n",
       "20  83290729     8         12        1106      \n",
       "21  95587716    35         21        5722      \n",
       "22  95587716     3          2        5722      \n",
       "23  85731228     6          9        1365      \n",
       "24  85731228     4          4        1365      \n",
       "25 140809253    15         12        4718      \n",
       "26 140809253    30          2        4718      \n",
       "27 202840531     4          2        2694      \n",
       "28 202840531    14         16        2694      \n",
       "29 139496402     4          6        6340      \n",
       "30 139496402     2          3        6340      \n",
       "⋮  ⋮            ⋮          ⋮        ⋮          \n",
       "51 245864840     4          7       12019      \n",
       "52 245864840     2          5       12019      \n",
       "53  16124147    73         77        1662      \n",
       "54  16124147    75         61        1662      \n",
       "55 182894382    27         19        1007      \n",
       "56 182894382     2          4        1007      \n",
       "57 216606601     2          2        8774      \n",
       "58 216606601     3          2        8774      \n",
       "59 220387372    22         12        7531      \n",
       "60 220387372    23         31        7531      \n",
       "61 150943942    13          7       30486      \n",
       "62 150943942     2          2       30486      \n",
       "63 163538559    15          4        1603      \n",
       "64 163538559    16         20        1603      \n",
       "65 168822280     4          4        6617      \n",
       "66 168822280     2          2        6617      \n",
       "67   8762856    11         19        4511      \n",
       "68   8762856     7          8        4511      \n",
       "69  47868266     2          4        3806      \n",
       "70  47868266     2          3        3806      \n",
       "71  53601035    57         44       10450      \n",
       "72  53601035     9          5       10450      \n",
       "73  63468494     7         12        5907      \n",
       "74  63468494     2          3        5907      \n",
       "75   4331183    56         19         879      \n",
       "76   4331183    10          6         879      \n",
       "77  97463458     2          2       10886      \n",
       "78  97463458     2          2       10886      \n",
       "79    127117     3          2       10466      \n",
       "80    127117     3          2       10466      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(MuGerminal_annotated)\n",
    "#head(MuGerminal_annotated)\n",
    "\n",
    "old_table <- read.csv(\"MuSeq_table_final/Mu_single_GeneIds_gene_lengths_and_stock.csv\")\n",
    "nrow(old_table)\n",
    "head(old_table)\n",
    "\n",
    "\n",
    "old_table  %>%\n",
    "  filter(grepl('RNA', GeneID))\n",
    "\n",
    "\n",
    "diff_vec <- setdiff(MuGerminal_annotated_stocks$InsertionStart, old_table$Start)\n",
    "MuGerminal_annotated %>%\n",
    "  filter(InsertionStart %in% diff_vec)\n",
    "\n",
    "## apparently in the old approach I missed those insertions,\n",
    "## which were overlapping with the borders of the gene model\n",
    "## for this test set = 40..\n",
    "## well, doesn't matter now I gues - or could I rerun and update\n",
    "## probablyy not worth the work - although a master student could do it ;D\n",
    "\n",
    "#test %>%\n",
    "#  filter(grepl('B73', Chr.x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a5509f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb02d6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aa02cfb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "928b7c2d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "304639b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
